internals/src/macros.rs

```diff
@@ -362,3 +362,140 @@ macro_rules! impl_array_newtype {
         }
     };
 }
+
+/// Implements hex-based string-ish traits for byte-based newtypes.
+///
+/// - `fmt::Display` and `str::FromStr` (using lowercase hex)
+/// - `fmt::LowerHex` and `UpperHex`
+/// - `fmt::Debug` (using `LowerHex`)
+///
+/// As well as an inherent `from_hex` method.
+#[macro_export]
+macro_rules! impl_array_newtype_hex {
+    ($t:ident, $len:literal) => {
+        impl $t {
+            /// Constructs a new `Self` from a hex string.
+            pub fn from_hex(s: &str) -> Result<Self, hex::HexToArrayError> {
```

Comment:
This line here adds a pre-1.0 type (from `hex v0.x.y`) to the public API of the caller.

I am super nervous of this PR in general, because of how many times we have been burned by macros.